hassil/recognize.py

```diff
@@ -314,174 +341,175 @@ def _remove_skip_words(text: str, skip_words: Iterable[str]) -> str:
         if seq.type == SequenceType.ALTERNATIVE:
             # Any may match (words | in | alternative)
             # NOTE: [optional] = (optional | )
             for item in seq.items:
-                yield from match_expression(context, item)
+                yield from match_expression(settings, context, item)
+
+        elif seq.type == SequenceType.GROUP:
+            if seq.items:
+                # All must match (words in group)
+                group_contexts = [context]
+                for item in seq.items:
+                    # Next step
+                    group_contexts = [
+                        item_context
+                        for group_context in group_contexts
+                        for item_context in match_expression(
+                            settings, group_context, item
+                        )
+                    ]
+                    if not group_contexts:
+                        break
 
-        elif seq.type == SequenceType.GROUP and seq.items:
-            # All must match (words in group)
-            group_contexts = [context]
-            for item in seq.items:
-                # Next step
-                group_contexts = [
-                    item_context
-                    for group_context in group_contexts
-                    for item_context in match_expression(group_context, item)
-                ]
-                if not group_contexts:
-                    break
-
-            for group_context in group_contexts:
-                yield group_context
+                for group_context in group_contexts:
+                    yield group_context
         else:
             raise ValueError(f"Unexpected sequence type: {seq}")
 
     elif isinstance(expression, ListReference):
         # {list}
         list_ref: ListReference = expression
-        if (not context.slot_lists) or (list_ref.list_name not in context.slot_lists):
+        if (not settings.slot_lists) or (list_ref.list_name not in settings.slot_lists):
             raise MissingListError(f"Missing slot list {{{list_ref.list_name}}}")
 
-        slot_list = context.slot_lists[list_ref.list_name]
-        if isinstance(slot_list, TextSlotList) and context.text:
-            text_list: TextSlotList = slot_list
-            # Any value may match
-            for slot_value in text_list.values:
-                value_contexts = match_expression(
-                    MatchContext(
-                        # Copy over
-                        text=context.text,
-                        slot_lists=context.slot_lists,
-                        expansion_rules=context.expansion_rules,
-                        entities=context.entities,
-                        intent_context=context.intent_context,
-                    ),
-                    slot_value.text_in,
-                )
+        slot_list = settings.slot_lists[list_ref.list_name]
+        if isinstance(slot_list, TextSlotList):
+            if context.text:
```

Comment:
Merge if statement again

Reply:
The merged if fails under certain conditions. If `context.text` is empty, then it will fall through to the `ValueError`.